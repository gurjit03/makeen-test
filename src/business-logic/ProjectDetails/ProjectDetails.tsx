import React from 'react';
import ProjectDetailsUI from '../../components/ProjectDetails/ProjectDetails';
import useFetch from '../../hooks/useFetch';
import { API_URL } from '../../config/app';
import { Project } from '../../interfaces/Project.interface';

interface ProjectDetailsProps {
    projectId: string
}

const ProjectDetails: React.FC<ProjectDetailsProps> = (props) => {
    const { projectId } = props;
    const PROJECT_API_URL = `${API_URL}/projects/${projectId}`
    const { data: project, status } = useFetch<Project>(PROJECT_API_URL);

    if (status === 'fetching') return <span>LOADING...</span>
    return project ? <ProjectDetailsUI {...project} /> : null
}

export default ProjectDetails;