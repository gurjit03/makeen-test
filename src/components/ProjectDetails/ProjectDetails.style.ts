import styled from 'styled-components'
import { colors, fontSizes, verticalSpacing } from '../../config/jss'

export const Wrapper = styled.div`
	padding: ${verticalSpacing * 4}px;
`

export const ProjectStatus = styled.p`
	font-size: ${fontSizes.xSmall}px;
	margin-bottom: ${verticalSpacing * 2}px;
`

export const ProjectName = styled.h2`
	font-size: ${fontSizes.large}px;
	margin-bottom: ${verticalSpacing * 2}px;
`

export const ProjectDetailLabel = styled.h3`
	color: ${colors.grey};
	margin-bottom: ${verticalSpacing}px;
`

export const ProjectDetailGroup = styled.div`
	margin-bottom: ${verticalSpacing * 2}px;
`

export const ProjectDescription = styled.p`
	font-size: ${fontSizes.large}px;
	letter-spacing: 1px;
`

export const ProjectConsultants = styled.ul`
	margin: 0;
	padding: 0;
`
