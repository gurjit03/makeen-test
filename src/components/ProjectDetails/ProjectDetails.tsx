import React from 'react'
import Consultant from '../Consultant/Consultant'
import { Project as ProjectDetailsProps } from '../../interfaces/Project.interface'
import { getColorsForStatus, getProperDate } from '../../utils'

import {
	ProjectDuration,
	ProjectEndDuration,
	ProjectStatus,
} from '../ProjectCard/ProjectCard.style'
import {
	ProjectDescription,
	ProjectDetailGroup,
	ProjectDetailLabel,
	ProjectName,
	ProjectConsultants,
	Wrapper,
} from '../ProjectDetails/ProjectDetails.style'

const ProjectDetails: React.FC<ProjectDetailsProps> = props => {
	const { name, status, duration, description, consultants = [] } = props
	const projectStatusColor = React.useMemo(
		() => getColorsForStatus(status),
		[status]
	)
	const projectStartDate = React.useMemo(
		() => getProperDate(duration.start),
		[duration, duration.start]
	)
	const projectEndDate = React.useMemo(
		() => getProperDate(duration.end),
		[duration, duration.end]
	)
	const allConsultants = React.useMemo(
		() =>
			consultants.map(consultant => (
				<Consultant key={consultant.id} {...consultant} />
			)),
		[consultants]
	)
	return (
		<Wrapper>
			<ProjectStatus color={projectStatusColor}>{status}</ProjectStatus>
			<ProjectName>{name}</ProjectName>
			<ProjectDetailGroup>
				<ProjectDetailLabel>Duration</ProjectDetailLabel>
				<ProjectDuration>
					<p>{projectStartDate}</p> to{' '}
					<ProjectEndDuration>{projectEndDate}</ProjectEndDuration>
				</ProjectDuration>
			</ProjectDetailGroup>
			<ProjectDetailGroup>
				<ProjectDetailLabel>Description</ProjectDetailLabel>
				<ProjectDescription>{description}</ProjectDescription>
			</ProjectDetailGroup>
			<ProjectDetailGroup>
				<ProjectDetailLabel>Consultants</ProjectDetailLabel>
				<ProjectConsultants>{allConsultants}</ProjectConsultants>
			</ProjectDetailGroup>
		</Wrapper>
	)
}

export default ProjectDetails
