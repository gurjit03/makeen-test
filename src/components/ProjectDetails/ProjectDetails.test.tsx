import React from 'react';
import { render, screen } from '@testing-library/react';
import projectMock from '../../mock/project.mock';
import ProjectDetails from './ProjectDetails'

describe('<ProjectDetails />', () => {
    it('should render properly', () => {
        const { asFragment } = render(<ProjectDetails {...projectMock} />)
        expect(asFragment()).toMatchSnapshot();
    });

    it('should display the consultants properly', () => {
        render(<ProjectDetails {...projectMock} />)
        const consultantName1 = screen.getByText('john doe1');
        expect(consultantName1).toBeInTheDocument()
        const consultantName2 = screen.getByText('john doe2');
        expect(consultantName2).toBeInTheDocument()
    })
})

