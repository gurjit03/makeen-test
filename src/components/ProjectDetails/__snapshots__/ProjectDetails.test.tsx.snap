// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectDetails /> should render properly 1`] = `
<DocumentFragment>
  <div
    class="sc-jrAFXE ewGNMM"
  >
    <p
      class="sc-gKseQn beAyqj"
      color="#3CA6E6"
    >
      running
    </p>
    <h2
      class="sc-iqHYmW bjKVDm"
    >
      john doe
    </h2>
    <div
      class="sc-dQoVA iRPZOA"
    >
      <h3
        class="sc-crrszt jRwIeB"
      >
        Duration
      </h3>
      <div
        class="sc-eCstlR ejPDsS"
      >
        <p>
          11.9.21
        </p>
         to 
        <p
          class="sc-fubCzh dcoLlf"
        >
          11.9.21
        </p>
      </div>
    </div>
    <div
      class="sc-dQoVA iRPZOA"
    >
      <h3
        class="sc-crrszt jRwIeB"
      >
        Description
      </h3>
      <p
        class="sc-bqyKOL fwwEzA"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>
    </div>
    <div
      class="sc-dQoVA iRPZOA"
    >
      <h3
        class="sc-crrszt jRwIeB"
      >
        Consultants
      </h3>
      <ul
        class="sc-kstqJO jRmbKI"
      >
        <li
          class="sc-bdfBQB dSprCg"
        >
          <img
            class="sc-gsTEea dtNHVy"
            src="https://avataaars.io/?avatarStyle=Circle&topType=LongHairStraight&accessoriesType=Blank&hairColor=BrownDark&facialHairType=Blank&clotheType=BlazerShirt&eyeType=Default&eyebrowType=Default&mouthType=Default&skinColor=Light'"
          />
          <p
            class="sc-dlfnuX iHPnGy"
          >
            john doe1
          </p>
        </li>
        <li
          class="sc-bdfBQB dSprCg"
        >
          <img
            class="sc-gsTEea dtNHVy"
            src="https://avataaars.io/?avatarStyle=Circle&topType=LongHairStraight&accessoriesType=Blank&hairColor=BrownDark&facialHairType=Blank&clotheType=BlazerShirt&eyeType=Default&eyebrowType=Default&mouthType=Default&skinColor=Light'"
          />
          <p
            class="sc-dlfnuX iHPnGy"
          >
            john doe2
          </p>
        </li>
        <li
          class="sc-bdfBQB dSprCg"
        >
          <img
            class="sc-gsTEea dtNHVy"
            src="https://avataaars.io/?avatarStyle=Circle&topType=LongHairStraight&accessoriesType=Blank&hairColor=BrownDark&facialHairType=Blank&clotheType=BlazerShirt&eyeType=Default&eyebrowType=Default&mouthType=Default&skinColor=Light'"
          />
          <p
            class="sc-dlfnuX iHPnGy"
          >
            john doe3
          </p>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
